api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,648 @@
+	"github.com/quintilesims/layer0/common/config"
+	"github.com/quintilesims/layer0/common/errors"
+	"github.com/quintilesims/layer0/common/models"
+)
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         *cache.Cache
+	environmentCache    *cache.Cache
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		deployCache:         cache.New(),
+		environmentCache:    cache.New(),
+	}
 }
 
+type ProviderDistribution struct {
+	Errs      []error
+	Providers []*scaler.ResourceProvider
+}
+
+// Scale determines whether or not instances need to be added to or removed from a Layer0 environment, and makes any necessary changes.
+// It consists of three primary logical groupings:
+// 1. Gather all providers (instances) and consumers (tasks/services) of resources.
+// 2. Calculate the optimal distribution of consumers among providers, including whether instances should be added or removed.
+// 3. Update the AutoScaling Group to realize the changes calculated previously.
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceConsumers, resourceProviders, err := e.GetCurrentState(clusterName)
+	if err != nil {
+		return err
+	}
+
+	var incompatibilityError error
+	resourceProviders, unusedProviders, desiredScale, err := e.CalculateOptimizedState(clusterName, resourceConsumers, resourceProviders)
+	if err != nil {
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.IncompatibleConsumerAndProvider {
+			incompatibilityError = err
+		} else {
+			return err
+		}
+	}
+
+	if err := e.ScaleToState(clusterName, desiredScale, unusedProviders); err != nil {
+		return err
+	}
+
+	return incompatibilityError
+}
+
+func (e *EnvironmentScaler) GetCurrentState(clusterName string) ([]scaler.ResourceConsumer, []*scaler.ResourceProvider, error) {
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceProviders for env '%s': %v", clusterName, resourceProviders)
+
+	var resourceConsumers []scaler.ResourceConsumer
+
+	serviceConsumers, err := e.getServiceResourceConsumers(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
```

Comment:
In an instance where my api instance was scaled at 0:0:0 and another env was scaled at 1:2:3, it seems that at some point after the resource providers are printed to the console, I get a `Family contains invalid characters error` back from ECS:

```2017/12/14 15:09:46 [INFO] [Worker] [2]: Starting job 1513292982109821621
2017/12/14 15:09:47 [INFO] [Worker] [1]: Starting job 1513292982607578204
2017/12/14 15:09:47 [DEBUG] [EnvironmentScaler] resourceProviders for env 'l0-jparsonsdev-api': []
2017/12/14 15:09:50 [DEBUG] [EnvironmentScaler] resourceProviders for env 'l0-jparsonsdev-scalertd6f75': [&scaler.ResourceProvider{AgentIsConnected:true, AvailableCPU:1024, AvailableMemory:1.04333312e+09, ID:"i-0ebd1ee81c6882fc0", InUse:false, Status:"ACTIVE", UsedPorts:[]int{22, 2376, 2375, 51678, 51679}} &scaler.ResourceProvider{AgentIsConnected:true, AvailableCPU:1024, AvailableMemory:1.04333312e+09, ID:"i-09199f3638e73b999", InUse:false, Status:"ACTIVE", UsedPorts:[]int{22, 2376, 2375, 51678, 51679}}]
2017/12/14 15:09:51 [ERROR] [Worker] [2]: Failed to run job 1513292982109821621: ClientException: Family contains invalid characters.
	status code: 400, request id: de96d064-e123-11e7-b722-037652736123
2017/12/14 15:09:53 [DEBUG] [EnvironmentScaler] resourceConsumers for env 'l0-jparsonsdev-scalertd6f75': []scaler.ResourceConsumer(nil)
2017/12/14 15:09:54 [DEBUG] [EnvironmentScaler] Attempting to scale environment 'l0-jparsonsdev-scalertd6f75' from current scale of '2' to desired scale of '0'.
2017/12/14 15:09:54 [DEBUG] [EnvironmentScaler] Will not scale below minimum capacity of '1'. Aborting scaling action for environment 'l0-jparsonsdev-scalertd6f75'.
2017/12/14 15:09:55 [INFO] [Worker] [1]: Finished job 1513292982607578204
```

When I deleted l0-jparsonsdev-scalertd6f75 and scaled the api env to something non-zero (like 1:1:2), then I'm not getting this exception. It appears there is some sort of condition where this exception is thrown if there are no services or deploys to pull from the cluster.

Reply:
From what I've seen, I can confirm that this error occurs when the `api` environment is scaled to `0:0:0`. At the moment, I'm not certain whether that's something in the scaler code, or a result of the `api` environment having been created through an older version of Layer0. I'll keep digging.

Reply:
I've got some more to report.

After dropping in some debugging print statements (and while running the API locally, of course), I've found that the cluster name doesn't function properly when the `api` environment is scaled to zero.

Some output when there exists at least one running instance for the `api`:
```
2017/12/14 19:43:47 [DEBUG] [ScalerDaemon]: Starting Run
2017/12/14 19:43:47 [INFO] [Worker] [2]: Starting job 1513309427736659375
2017/12/14 19:43:47 [INFO] [Worker] [1]: Starting job 1513309427758736000
2017/12/14 19:43:48 [DEBUG] [EnvironmentScaler] Attempting to scale environment 'l0-tlakedev-testa80c9bd4' from current scale of '1' to desired scale of '0'.
2017/12/14 19:43:48 [DEBUG] [EnvironmentScaler] Will not scale below minimum capacity of '1'. Aborting scaling action for environment 'l0-tlakedev-testa80c9bd4'.
2017/12/14 19:43:48 [DEBUG] [EnvironmentScaler] Attempting to scale environment 'l0-tlakedev-api' from current scale of '2' to desired scale of '1'.
2017/12/14 19:43:48 [DEBUG] [EnvironmentScaler] Will not scale below minimum capacity of '2'. Aborting scaling action for environment 'l0-tlakedev-api'.
2017/12/14 19:43:48 [INFO] [Worker] [1]: Finished job 1513309427758736000
2017/12/14 19:43:48 [INFO] [Worker] [2]: Finished job 1513309427736659375
```

Some output when the `api` environment has been scaled to zero:
```
2017/12/14 19:43:37 [DEBUG] [ScalerDaemon]: Starting Run
2017/12/14 19:43:37 [DEBUG] [ScalerDaemon] Creating scale job for environment api
2017/12/14 19:43:37 [DEBUG] [ScalerDaemon] Creating scale job for environment testa80c9bd4
2017/12/14 19:43:37 [INFO] [Worker] [2]: Starting job 1513309417745904574
2017/12/14 19:43:37 [INFO] [Worker] [1]: Starting job 1513309417770146400
2017/12/14 19:43:38 [DEBUG] [HUNTER] In getServiceResourceConsumers(): Obtaining container resources using deployID '"ecs-svc/9223370532215438139"'
2017/12/14 19:43:38 [DEBUG] [HUNTER] No consumers found in deploy cache for deployID '"ecs-svc/9223370532215438139"'.
2017/12/14 19:43:38 [DEBUG] [HUNTER] Err from ECS.DescribeTaskDefinition using input with deployID 'ecs-svc/9223370532215438139' (error: &awserr.requestError{awsError:(*awserr.baseError)(0xc4201ebd00), statusCode:400, requestID:"d8a40784-e149-11e7-bc82-8b9df38ac751"})
2017/12/14 19:43:38 [DEBUG] [EnvironmentScaler] Attempting to scale environment 'l0-tlakedev-testa80c9bd4' from current scale of '1' to desired scale of '0'.
2017/12/14 19:43:46 [DEBUG] [EnvironmentScaler] Will not scale below minimum capacity of '1'. Aborting scaling action for environment 'l0-tlakedev-testa80c9bd4'.
2017/12/14 19:43:46 [ERROR] [Worker] [2]: Failed to run job 1513309417745904574: ClientException: Family contains invalid characters.
        status code: 400, request id: d8a40784-e149-11e7-bc82-8b9df38ac751
2017/12/14 19:43:46 [INFO] [Worker] [1]: Finished job 1513309417770146400
```

Note that in the second scenario (when `api` is scaled to zero), Layer0 uses a _different cluster name:_ `"ecs-svc/9223370532215438139"`, not the `"l0-tlakedev-api"` that I would expect. We can see that Layer0 eventually makes a call to `ECS.DescribeTaskDefinition()` using that same `ecs-svc` string for the `deployID`.

(The invalid character it's referring to in the error message is the `/`.)

_If this only happens_ when the `api` environment is scaled down to zero (i.e., this error does not arise when a non-`api` environment is scaled to zero but still wants to deploy a service upon it), I'm not sure how much of a big deal this bug is; in theory, a user will never be running the Layer0 API locally, in which case, scaling down the `api` environment would cause problems anyway because that's the API that their client would be trying (and failing) to communicate with.

Reply:
Another update: After setting up _walkthrough-1_ complete with one service, I scaled the `demo-env` environment down to `0:0:0`. Of course, there was still a service looking to be placed into an instance, and I was able to replicate the same error for this setup. In other words, this error is not unique to the `api` environment. _Any time_ there exists an environment scaled to `0:0:0` and at least one service looking to be placed into that environment, this 'invalid character' error will pop up as Layer0 tries to gather resources using some string like `ecs-svc/1234567890` instead of one our safe, slash-less strings.

Reply:
gh-

Reply:
Oh cool and another thing I just found.

Apparently, something will muss with the scaling of an environment. In my work to dial in this error, I've scaled the guestbook environment (`demo-env`) all the way down to `0:0:0`. Then, I let the local API run for a while and also set up another monitoring bash loop to run `l0 environment get demo-env` at ten-second intervals. At one point during these runs, the `demo-env`'s scale had changed to `0:1:1` without any intervention from myself. Nothing in the API's debug statements seemed to reflect any change, either.

From the monitoring bash loop:
```
Thu Dec 14 21:04:01 PST 2017
ENVIRONMENT ID  ENVIRONMENT NAME  OS     SCALE  INSTANCE TYPE  LINKS
demoenvda478    demo-env          linux  0:0:0  t2.small

Thu Dec 14 21:04:15 PST 2017
ENVIRONMENT ID  ENVIRONMENT NAME  OS     SCALE  INSTANCE TYPE  LINKS
demoenvda478    demo-env          linux  0:1:1  t2.small
```

We can see that the event must have happened between `21:04:01` and `21:04:15`, and so here are the API logs from that timebox:
```
2017/12/14 21:04:02 GET /tag?fuzz=demo-env&type=environment
2017/12/14 21:04:05 GET /environment/demoenvda478
2017/12/14 21:04:06 [DEBUG] [ScalerDaemon]: Starting Run
2017/12/14 21:04:06 [DEBUG] [ScalerDaemon] Creating scale job for environment api
2017/12/14 21:04:06 [DEBUG] [ScalerDaemon] Creating scale job for environment demoenvda478
2017/12/14 21:04:07 [INFO] [Worker] [1]: Starting job 1513314246970333294
2017/12/14 21:04:07 [INFO] [Worker] [3]: Starting job 1513314246995481029
2017/12/14 21:04:07 [DEBUG] [HUNTER] In getServiceResourceConsumers(): Obtaining container resources using deployID '"ecs-svc/9223370523544020128"'
2017/12/14 21:04:07 [DEBUG] [HUNTER] At top of getContainerResourceFromDeploy(): deployID passed in: '"ecs-svc/9223370523544020128"'
2017/12/14 21:04:07 [DEBUG] [HUNTER] No consumers found in deploy cache for deployID '"ecs-svc/9223370523544020128"'.
2017/12/14 21:04:07 [DEBUG] [HUNTER] ECS.DescribeTaskDefinition using input with deployID 'ecs-svc/9223370523544020128'
2017/12/14 21:04:07 [DEBUG] [HUNTER] Err from ECS.DescribeTaskDefinition using input with deployID 'ecs-svc/9223370523544020128' (error: &awserr.requestError{awsError:(*awserr.baseError)(0xc42036b440), statusCode:400, requestID:"171ee619-e155-11e7-ae64-974b3cfc82d9"})
2017/12/14 21:04:07 [ERROR] [Worker] [3]: Failed to run job 1513314246995481029: ClientException: Family contains invalid characters.
        status code: 400, request id: 171ee619-e155-11e7-ae64-974b3cfc82d9
2017/12/14 21:04:07 [DEBUG] [EnvironmentScaler] Attempting to scale environment 'l0-tlakedev-api' from current scale of '2' to desired scale of '1'.
2017/12/14 21:04:07 [DEBUG] [EnvironmentScaler] Will not scale below minimum capacity of '2'. Aborting scaling action for environment 'l0-tlakedev-api'.
2017/12/14 21:04:07 [INFO] [Worker] [1]: Finished job 1513314246970333294
2017/12/14 21:04:15 GET /tag?fuzz=demo-env&type=environment
2017/12/14 21:04:15 GET /environment/demoenvda478
```

The logs are the same both before and after the timebox as well.

Reply:
This is now issue #449.